pkg/controller/bastion/actuator_delete.go

```diff
@@ -0,0 +1,199 @@
+	if err != nil {
+		return fmt.Errorf("failed to remove nsg rules: %w", err)
+	}
+
+	return nil
+}
+
+func removeNSGRule(ctx context.Context, factory azureclient.Factory, logger logr.Logger, opt *Options) error {
+	securityGroupResp, err := getNetworkSecurityGroup(ctx, factory, opt)
+	if err != nil {
+		return err
+	}
+
+	ruleList := []string{NSGIngressAllowSSHResourceName(opt.BastionInstanceName), NSGEgressDenyAllResourceName(opt.BastionInstanceName), NSGEgressAllowOnlyResourceName(opt.BastionInstanceName)}
+	for _, name := range ruleList {
+		*securityGroupResp.SecurityRules = removeSecurityRuleBody(name, *securityGroupResp.SecurityRules)
+	}
+
+	err = createOrUpdateNetworkSecGroup(ctx, factory, opt, securityGroupResp)
+	if err != nil {
+		return err
+	}
+
+	logger.Info("bastion network security group rules removed: ", "nsg", NSGIngressAllowSSHResourceName(opt.BastionInstanceName), NSGEgressDenyAllResourceName(opt.BastionInstanceName), NSGEgressAllowOnlyResourceName(opt.BastionInstanceName))
+	return nil
+}
+
+func removePublicIP(ctx context.Context, factory azureclient.Factory, logger logr.Logger, opt *Options) error {
+	ip, err := getPublicIP(ctx, factory, opt)
+	if err != nil {
+		return err
+	}
+
+	if ip == nil {
+		return nil
+	}
+
+	publicClient, err := factory.NetworkPublicIP(ctx, opt.SecretReference)
+	if err != nil {
+		return err
+	}
+
+	err = publicClient.Delete(ctx, opt.ResourceGroupName, opt.BastionPublicIPName)
+	if err != nil {
+		return fmt.Errorf("failed to delete Public IP: %w", err)
+	}
+
+	logger.Info("Public IP removed", "ip", opt.BastionPublicIPName)
+	return nil
+}
+
+func removeNic(ctx context.Context, factory azureclient.Factory, logger logr.Logger, opt *Options) error {
+	nic, err := getNic(ctx, factory, opt)
+	if err != nil {
+		return err
+	}
+
+	if nic == nil {
+		return nil
+	}
+
+	nicClient, err := factory.NetworkInterface(ctx, opt.SecretReference)
+	if err != nil {
+		return err
+	}
+
+	err = nicClient.Delete(ctx, opt.ResourceGroupName, opt.NicName)
+	if err != nil {
+		return fmt.Errorf("failed to delete Nic: %w", err)
+	}
+
+	logger.Info("Nic removed", "nic", opt.NicName)
+	return nil
+}
+
+func removeDisk(ctx context.Context, factory azureclient.Factory, logger logr.Logger, opt *Options) error {
+	disk, err := getDisk(ctx, factory, opt)
+	if err != nil {
+		return err
+	}
+	if disk == nil {
+		return nil
+	}
+	diskClient, err := factory.Disk(ctx, opt.SecretReference)
+	if err != nil {
+		return err
+	}
+	err = diskClient.Delete(ctx, opt.ResourceGroupName, opt.DiskName)
+	if err != nil {
+		return fmt.Errorf("failed to delete disk: %w", err)
+	}
+
+	logger.Info("Disk removed", "disk", opt.DiskName)
+	return nil
+}
+
+func removeBastionInstance(ctx context.Context, factory azureclient.Factory, logger logr.Logger, opt *Options) error {
+	instance, err := getBastionInstance(ctx, factory, opt)
+	if err != nil {
+		return err
+	}
+	if instance == nil {
+		return nil
+	}
```

Comment:
Could we not simply try do delete the machine and return nil in case we get an not found error?
Then we would not need to fetch it upfront.

Reply:
good idea, ecf7ca13